File: src/lib/builds/equipReady.test.ts
import { describe, expect, it } from "vitest";

import type { ResolvedVariantEquipment } from "@/lib/builds/resolveVariant";
import {
  assertEquipReady,
  buildInventoryPinIndex,
  computeEquipReady,
} from "@/lib/builds/equipReady";
import { API_ERROR_CODES } from "@/lib/api/errors";

function resolved(
  equipment: ResolvedVariantEquipment["equipment"],
): ResolvedVariantEquipment {
  return { equipment, conflicts: [] };
}

describe("computeEquipReady", () => {
  it("marks slots without instanceId as wishlist and not equip-ready", () => {
    const result = computeEquipReady(
      resolved({
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
        },
      }),
      buildInventoryPinIndex([]),
    );
    expect(result.equipReady).toBe(false);
    expect(result.pinStatuses).toEqual([{ slot: "primary", status: "wishlist" }]);
  });

  it("is equip-ready when every applied combat slot is pinned", () => {
    const inventory = buildInventoryPinIndex([
      { instanceId: "a", itemHash: 1 },
      { instanceId: "b", itemHash: 2 },
    ]);
    const result = computeEquipReady(
      resolved({
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
          instanceId: "a",
        },
        helmet: {
          slot: "helmet",
          itemHash: 2,
          itemName: "Helm",
          source: "set",
          instanceId: "b",
        },
      }),
      inventory,
    );
    expect(result.equipReady).toBe(true);
    expect(result.pinStatuses.every((s) => s.status === "pinned")).toBe(true);
  });

  it("ignores empty non-default gaps (only applied slots)", () => {
    const inventory = buildInventoryPinIndex([{ instanceId: "a", itemHash: 1 }]);
    const result = computeEquipReady(
      resolved({
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
          instanceId: "a",
        },
      }),
      inventory,
    );
    expect(result.equipReady).toBe(true);
    expect(result.pinStatuses).toHaveLength(1);
  });

  it("marks missing inventory instance as stale", () => {
    const result = computeEquipReady(
      resolved({
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
          instanceId: "gone",
        },
      }),
      buildInventoryPinIndex([]),
    );
    expect(result.equipReady).toBe(false);
    expect(result.pinStatuses[0]).toMatchObject({
      status: "stale",
      reason: "instance_missing",
    });
  });

  it("assertEquipReady throws NOT_EQUIP_READY when not ready", () => {
    const result = computeEquipReady(
      resolved({
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
        },
      }),
      buildInventoryPinIndex([]),
    );
    expect(() => assertEquipReady(result)).toThrow(
      expect.objectContaining({ code: API_ERROR_CODES.NOT_EQUIP_READY }),
    );
  });

  it("post-sync: pre-ready pins become not equip-ready when instance missing after refresh", () => {
    const equipment = {
      primary: {
        slot: "primary" as const,
        itemHash: 1,
        itemName: "Gun",
        source: "set" as const,
        instanceId: "pin-a",
      },
      helmet: {
        slot: "helmet" as const,
        itemHash: 2,
        itemName: "Helm",
        source: "set" as const,
        instanceId: "pin-b",
      },
    };

    const preSync = computeEquipReady(
      resolved(equipment),
      buildInventoryPinIndex([
        { instanceId: "pin-a", itemHash: 1 },
        { instanceId: "pin-b", itemHash: 2 },
      ]),
    );
    expect(preSync.equipReady).toBe(true);

    // Post-sync inventory lost pin-b (sold/transferred out of account view).
    const postSync = computeEquipReady(
      resolved(equipment),
      buildInventoryPinIndex([{ instanceId: "pin-a", itemHash: 1 }]),
    );
    expect(postSync.equipReady).toBe(false);
    expect(postSync.pinStatuses).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          slot: "helmet",
          status: "stale",
          reason: "instance_missing",
          instanceId: "pin-b",
        }),
      ]),
    );
    expect(() => assertEquipReady(postSync)).toThrow(
      expect.objectContaining({
        code: API_ERROR_CODES.NOT_EQUIP_READY,
        status: 409,
      }),
    );
  });

  it("post-sync happy path: all claimed instances still present remain equip-ready", () => {
    const equipment = {
      primary: {
        slot: "primary" as const,
        itemHash: 1,
        itemName: "Gun",
        source: "set" as const,
        instanceId: "pin-a",
      },
      special: {
        slot: "special" as const,
        itemHash: 3,
        itemName: "Special",
        source: "set" as const,
        instanceId: "pin-c",
      },
      helmet: {
        slot: "helmet" as const,
        itemHash: 2,
        itemName: "Helm",
        source: "set" as const,
        instanceId: "pin-b",
      },
    };
    const postSync = computeEquipReady(
      resolved(equipment),
      buildInventoryPinIndex([
        { instanceId: "pin-a", itemHash: 1 },
        { instanceId: "pin-b", itemHash: 2 },
        { instanceId: "pin-c", itemHash: 3 },
      ]),
    );
    expect(postSync.equipReady).toBe(true);
    expect(() => assertEquipReady(postSync)).not.toThrow();
  });
});

File: flutter/packages/domain/test/equip_ready_test.dart
import 'package:destiny2_domain/destiny2_domain.dart';
import 'package:test/test.dart';

ResolvedVariantEquipment resolved(
  Map<EquipmentSlot, SlotClaim> equipment,
) {
  return ResolvedVariantEquipment(equipment: equipment, conflicts: const []);
}

SlotClaim claim(
  EquipmentSlot slot,
  int itemHash, {
  String? instanceId,
  String itemName = 'Item',
}) {
  return SlotClaim(
    slot: slot,
    itemHash: itemHash,
    itemName: itemName,
    source: ClaimSource.set,
    instanceId: instanceId,
  );
}

void main() {
  group('computeEquipReady', () {
    test('marks slots without instanceId as wishlist and not equip-ready', () {
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(EquipmentSlot.primary, 1, itemName: 'Gun'),
        }),
        buildInventoryPinIndex(const []),
      );
      expect(result.equipReady, isFalse);
      expect(result.pinStatuses, [
        const PinStatus(
          slot: EquipmentSlot.primary,
          status: PinStatusKind.wishlist,
        ),
      ]);
    });

    test('is equip-ready when every applied combat slot is pinned', () {
      final inventory = buildInventoryPinIndex(const [
        InventoryPinItem(instanceId: 'a', itemHash: 1),
        InventoryPinItem(instanceId: 'b', itemHash: 2),
      ]);
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'a',
            itemName: 'Gun',
          ),
          EquipmentSlot.helmet: claim(
            EquipmentSlot.helmet,
            2,
            instanceId: 'b',
            itemName: 'Helm',
          ),
        }),
        inventory,
      );
      expect(result.equipReady, isTrue);
      expect(
        result.pinStatuses.every((s) => s.status == PinStatusKind.pinned),
        isTrue,
      );
    });

    test('ignores empty non-default gaps (only applied slots)', () {
      final inventory = buildInventoryPinIndex(const [
        InventoryPinItem(instanceId: 'a', itemHash: 1),
      ]);
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'a',
            itemName: 'Gun',
          ),
        }),
        inventory,
      );
      expect(result.equipReady, isTrue);
      expect(result.pinStatuses, hasLength(1));
    });

    test('marks missing inventory instance as stale', () {
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'gone',
            itemName: 'Gun',
          ),
        }),
        buildInventoryPinIndex(const []),
      );
      expect(result.equipReady, isFalse);
      expect(result.pinStatuses.single.status, PinStatusKind.stale);
      expect(result.pinStatuses.single.reason, PinStaleReason.instanceMissing);
      expect(result.pinStatuses.single.reason!.wireName, 'instance_missing');
      expect(result.pinStatuses.single.instanceId, 'gone');
    });

    test('marks hash mismatch as stale', () {
      final inventory = buildInventoryPinIndex(const [
        InventoryPinItem(instanceId: 'a', itemHash: 999),
      ]);
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'a',
            itemName: 'Gun',
          ),
        }),
        inventory,
      );
      expect(result.equipReady, isFalse);
      expect(result.pinStatuses.single, const PinStatus(
        slot: EquipmentSlot.primary,
        status: PinStatusKind.stale,
        instanceId: 'a',
        reason: PinStaleReason.hashMismatch,
      ));
      expect(
        result.pinStatuses.single.reason!.wireName,
        'hash_mismatch',
      );
    });

    test('empty equipment is never equip-ready', () {
      final result = computeEquipReady(
        const ResolvedVariantEquipment(),
        buildInventoryPinIndex(const []),
      );
      expect(result.equipReady, isFalse);
      expect(result.pinStatuses, isEmpty);
    });

    test('mixed wishlist and pinned is not equip-ready', () {
      final inventory = buildInventoryPinIndex(const [
        InventoryPinItem(instanceId: 'a', itemHash: 1),
      ]);
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'a',
            itemName: 'Gun',
          ),
          EquipmentSlot.helmet: claim(
            EquipmentSlot.helmet,
            2,
            itemName: 'Helm',
          ),
        }),
        inventory,
      );
      expect(result.equipReady, isFalse);
      expect(
        result.pinStatuses.map((s) => s.status).toList(),
        [PinStatusKind.pinned, PinStatusKind.wishlist],
      );
    });

    test('assertEquipReady throws NOT_EQUIP_READY when not ready', () {
      final result = computeEquipReady(
        resolved({
          EquipmentSlot.primary: claim(EquipmentSlot.primary, 1, itemName: 'Gun'),
        }),
        buildInventoryPinIndex(const []),
      );
      expect(
        () => assertEquipReady(result),
        throwsA(
          isA<EquipReadyException>()
              .having((e) => e.code, 'code', DomainFailureCodes.notEquipReady)
              .having((e) => e.code, 'code string', 'NOT_EQUIP_READY')
              .having((e) => e.details?['allowed'], 'allowed', false),
        ),
      );
    });

    test(
      'post-sync: pre-ready pins become not equip-ready when instance missing after refresh',
      () {
        final equipment = {
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'pin-a',
            itemName: 'Gun',
          ),
          EquipmentSlot.helmet: claim(
            EquipmentSlot.helmet,
            2,
            instanceId: 'pin-b',
            itemName: 'Helm',
          ),
        };

        final preSync = computeEquipReady(
          resolved(equipment),
          buildInventoryPinIndex(const [
            InventoryPinItem(instanceId: 'pin-a', itemHash: 1),
            InventoryPinItem(instanceId: 'pin-b', itemHash: 2),
          ]),
        );
        expect(preSync.equipReady, isTrue);

        // Post-sync inventory lost pin-b (sold/transferred out of account view).
        final postSync = computeEquipReady(
          resolved(equipment),
          buildInventoryPinIndex(const [
            InventoryPinItem(instanceId: 'pin-a', itemHash: 1),
          ]),
        );
        expect(postSync.equipReady, isFalse);
        expect(
          postSync.pinStatuses,
          contains(
            const PinStatus(
              slot: EquipmentSlot.helmet,
              status: PinStatusKind.stale,
              instanceId: 'pin-b',
              reason: PinStaleReason.instanceMissing,
            ),
          ),
        );
        expect(
          () => assertEquipReady(postSync),
          throwsA(
            isA<EquipReadyException>().having(
              (e) => e.code,
              'code',
              DomainFailureCodes.notEquipReady,
            ),
          ),
        );
      },
    );

    test(
      'post-sync happy path: all claimed instances still present remain equip-ready',
      () {
        final equipment = {
          EquipmentSlot.primary: claim(
            EquipmentSlot.primary,
            1,
            instanceId: 'pin-a',
            itemName: 'Gun',
          ),
          EquipmentSlot.special: claim(
            EquipmentSlot.special,
            3,
            instanceId: 'pin-c',
            itemName: 'Special',
          ),
          EquipmentSlot.helmet: claim(
            EquipmentSlot.helmet,
            2,
            instanceId: 'pin-b',
            itemName: 'Helm',
          ),
        };
        final postSync = computeEquipReady(
          resolved(equipment),
          buildInventoryPinIndex(const [
            InventoryPinItem(instanceId: 'pin-a', itemHash: 1),
            InventoryPinItem(instanceId: 'pin-b', itemHash: 2),
            InventoryPinItem(instanceId: 'pin-c', itemHash: 3),
          ]),
        );
        expect(postSync.equipReady, isTrue);
        expect(() => assertEquipReady(postSync), returnsNormally);
      },
    );

    test('buildInventoryPinIndex maps instanceId to itemHash', () {
      final index = buildInventoryPinIndex(const [
        InventoryPinItem(instanceId: 'x', itemHash: 10),
        InventoryPinItem(instanceId: 'y', itemHash: 20),
      ]);
      expect(index['x'], 10);
      expect(index['y'], 20);
      expect(index.containsKey('z'), isFalse);
    });
  });
}
